libdnf/repo/Repo.cpp

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * Licensed under the GNU Lesser General Public License Version 2.1
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#define ASCII_LOWERCASE "abcdefghijklmnopqrstuvwxyz"
+#define ASCII_UPPERCASE "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+#define ASCII_LETTERS ASCII_LOWERCASE ASCII_UPPERCASE
+#define DIGITS "0123456789"
+#define REPOID_CHARS ASCII_LETTERS DIGITS "-_.:"
+
+#include "Repo.hpp"
+#include "../dnf-utils.h"
+#include "../hy-iutil.h"
+#include "../hy-util-private.hpp"
+#include "../hy-types.h"
+
+#include <librepo/librepo.h>
+#include <solv/repo.h>
+#include <utime.h>
+
+#include <map>
+#include <glib.h>
+
+namespace libdnf {
+
+class Repo::Impl {
+public:
+    Impl(const std::string & id, std::unique_ptr<ConfigRepo> && config);
+    ~Impl();
+
+    bool load();
+    bool loadCache();
+    bool canReuse();
+    void fetch();
+    std::string getCachedir();
+    char ** getMirrors();
+    int getAge() const;
+    void expire();
+    bool expired() const;
+    int expiresIn();
+    void resetTimestamp();
+    LrHandle * lrHandleInitBase();
+    LrHandle * lrHandleInitLocal();
+    LrHandle * lrHandleInitRemote(const char *destdir);
+
+    std::string id;
+    std::unique_ptr<ConfigRepo> conf;
+
+    ::Repo *libsolvRepo;
+    char ** mirrors;
+    // 0 forces expiration on the next call to load(), -1 means undefined value
+    int timestamp;
+    int max_timestamp;
+    std::string repomd_fn;
+    std::string primary_fn;
+    std::string filelists_fn;
+    std::string presto_fn;
+    std::string updateinfo_fn;
+    std::string comps_fn;
+    std::string revision;
+    GSList * content_tags;
```

Comment:
what is this?

Reply:
@ignatenkobrain what do you mean exactly?

Reply:
@ignatenkobrain if you were referring to `content_tags`: it's going to be a `vector` or something (surely not `GSList`), this is just temporary

Reply:
The question is "list of WHAT"? ;)

Reply:
Take a look here: http://rpm-software-management.github.io/librepo/lib.html#classes
It's used by DNF:
```
$ git grep _content_tags
dnf/cli/commands/repolist.py:                if md and md._content_tags:
dnf/cli/commands/repolist.py:                    tags = md._content_tags
dnf/repo.py:    def _content_tags(self):
```

Reply:
Basically, it's one of the items that librepo parses and returns in a dictionary. As for its purpose, I'm not really sure either (haven't researched yet). If it turns out redundant/unneeded/whatever, we will just remove it later :)